File: frontend/src/lib/roles.js
// Core roles form a fixed departmental hierarchy: Requestor -> Supervisor ->
// Manager -> Executive. Approval routing is resolved from per-department
// assignments (executive/manager/supervisor groups), not from role names.
export const ROLE_OPTIONS = [
  { value: "requestor", label: "Requestor" },
  { value: "both", label: "Requestor and Approver" },
  { value: "supervisor", label: "Supervisor" },
  { value: "manager", label: "Manager" },
  { value: "executive", label: "Executive" },
  { value: "super_admin", label: "Super Admin" },
];

// Legacy role values that may still exist on older user records. They behave
// as aliases of the core manager/executive roles.
const LEGACY_ROLE_LABELS = {
  approver: "Approver (Legacy)",
  manager_ops: "Manager (OPS) (Legacy)",
  manager_sup: "Manager (SUP) (Legacy)",
  executive_ops: "Executive (OPS) (Legacy)",
  executive_sup: "Executive (SUP) (Legacy)",
};

export const LEGACY_MANAGER_ROLES = ["manager_ops", "manager_sup"];
export const LEGACY_EXECUTIVE_ROLES = ["executive_ops", "executive_sup"];
export const MANAGER_ROLES = ["manager", ...LEGACY_MANAGER_ROLES];
export const EXECUTIVE_ROLES = ["executive", ...LEGACY_EXECUTIVE_ROLES];
export const FORM_MANAGER_ROLES = MANAGER_ROLES;
export const APPROVER_ROLES = [
  "approver",
  "both",
  "supervisor",
  "super_admin",
  ...MANAGER_ROLES,
  ...EXECUTIVE_ROLES,
];
export const REQUESTOR_ROLES = [
  "requestor",
  "both",
  "supervisor",
  "super_admin",
  ...MANAGER_ROLES,
  ...EXECUTIVE_ROLES,
];

export function isSuperAdminRole(role) {
  return role === "super_admin";
}

export function isManagerRole(role) {
  return MANAGER_ROLES.includes(role);
}

export function isExecutiveRole(role) {
  return EXECUTIVE_ROLES.includes(role);
}

export function isSupervisorRole(role) {
  return role === "supervisor";
}

export function isApproverRole(role) {
  return APPROVER_ROLES.includes(role);
}

export function isRequestorRole(role) {
  return REQUESTOR_ROLES.includes(role);
}

export function getRoleLabel(role) {
  return (
    ROLE_OPTIONS.find((option) => option.value === role)?.label ||
    LEGACY_ROLE_LABELS[role] ||
    role?.replace("_", " ") ||
    ""
  );
}

File: backend/utils/roles.py
# Core roles form a fixed departmental hierarchy. Who actually approves a
# request is resolved at request time from per-department assignments
# (department executive/manager/supervisor groups), not from role names.
REQUESTOR = "requestor"
BOTH = "both"  # Requestor and Approver
SUPERVISOR = "supervisor"
MANAGER = "manager"
EXECUTIVE = "executive"
SUPER_ADMIN = "super_admin"

# Legacy role values that may still exist on user documents. They keep
# working as aliases of the core manager/executive roles.
LEGACY_MANAGER_ROLES = {"manager_ops", "manager_sup"}
LEGACY_EXECUTIVE_ROLES = {"executive_ops", "executive_sup"}

MANAGER_ROLES = {MANAGER} | LEGACY_MANAGER_ROLES
EXECUTIVE_ROLES = {EXECUTIVE} | LEGACY_EXECUTIVE_ROLES

# Roles allowed to manage forms for their department.
FORM_MANAGER_ROLES = MANAGER_ROLES

REQUESTOR_ROLES = {REQUESTOR, BOTH, SUPERVISOR, SUPER_ADMIN} | MANAGER_ROLES | EXECUTIVE_ROLES
APPROVER_ROLES = {"approver", BOTH, SUPERVISOR, SUPER_ADMIN} | MANAGER_ROLES | EXECUTIVE_ROLES

# Managerial hierarchy levels used to resolve role-based approval steps.
# A head at level N approves requestors below level N.
SUPERVISOR_LEVEL = 1
MANAGER_LEVEL = 2
EXECUTIVE_LEVEL = 3


def hierarchy_level(role):
    """Managerial level of a role: 0 for non-managerial requestors/approvers."""
    if role == SUPERVISOR:
        return SUPERVISOR_LEVEL
    if role in MANAGER_ROLES:
        return MANAGER_LEVEL
    if role in EXECUTIVE_ROLES:
        return EXECUTIVE_LEVEL
    return 0


def role_of(user):
    return (user or {}).get("role", "")


def user_hierarchy_level(user):
    return hierarchy_level(role_of(user))


def is_super_admin(user):
    return role_of(user) == SUPER_ADMIN


def is_manager_role(role):
    return role in MANAGER_ROLES


def is_manager(user):
    return is_manager_role(role_of(user))


def is_executive_role(role):
    return role in EXECUTIVE_ROLES


def is_supervisor_role(role):
    return role == SUPERVISOR


def is_approval_capable(user):
    return role_of(user) in APPROVER_ROLES


def is_requestor_capable(user):
    return role_of(user) in REQUESTOR_ROLES
